// @nx/storybook/src/generators/init/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initSchematic = exports.initGenerator = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const js_1 = require("@nx/js");
const versions_1 = require("../../utils/versions");
const utilities_1 = require("../../utils/utilities");
const semver_1 = require("semver");
function checkDependenciesInstalled(host, schema) {
    var _a;
    const packageJson = (0, devkit_1.readJson)(host, 'package.json');
    const devDependencies = {};
    const dependencies = {};
    packageJson.dependencies = packageJson.dependencies || {};
    packageJson.devDependencices = packageJson.devDependencices || {};
    // base deps
    devDependencies['@nx/storybook'] = versions_1.nxVersion;
    let storybook7VersionToInstall = versions_1.storybookVersion;
    if ((0, utilities_1.storybookMajorVersion)() === 7 &&
        (0, utilities_1.getInstalledStorybookVersion)() &&
        (0, semver_1.gte)((0, utilities_1.getInstalledStorybookVersion)(), '7.0.0')) {
        storybook7VersionToInstall = (0, utilities_1.getInstalledStorybookVersion)();
    }
    // Needed for Storybook 7
    // https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#react-peer-dependencies-required
    if (!packageJson.dependencies['react'] &&
        !packageJson.devDependencies['react']) {
        dependencies['react'] = versions_1.reactVersion;
    }
    if (!packageJson.dependencies['react-dom'] &&
        !packageJson.devDependencies['react-dom']) {
        dependencies['react-dom'] = versions_1.reactVersion;
    }
    devDependencies['@storybook/core-server'] = storybook7VersionToInstall;
    devDependencies['@storybook/addon-essentials'] = storybook7VersionToInstall;
    if (schema.uiFramework) {
        if (schema.uiFramework === '@storybook/react-native') {
            devDependencies['@storybook/react-native'] = versions_1.storybookReactNativeVersion;
        }
        else {
            devDependencies[schema.uiFramework] = storybook7VersionToInstall;
            const isPnpm = (0, devkit_1.detectPackageManager)(host.root) === 'pnpm';
            if (isPnpm) {
                // If it's pnpm, it needs the framework without the builder
                // as a dependency too (eg. @storybook/react)
                const matchResult = (_a = schema.uiFramework) === null || _a === void 0 ? void 0 : _a.match(/^@storybook\/(\w+)/);
                const uiFrameworkWithoutBuilder = matchResult ? matchResult[0] : null;
                if (uiFrameworkWithoutBuilder) {
                    devDependencies[uiFrameworkWithoutBuilder] =
                        storybook7VersionToInstall;
                }
            }
        }
        if (schema.uiFramework === '@storybook/angular') {
            if (!packageJson.dependencies['@angular/forms'] &&
                !packageJson.devDependencies['@angular/forms']) {
                devDependencies['@angular/forms'] = '*';
            }
        }
        if (schema.uiFramework === '@storybook/web-components-vite' ||
            schema.uiFramework === '@storybook/web-components-webpack5') {
            devDependencies['lit'] = versions_1.litVersion;
        }
        if (schema.uiFramework === '@storybook/react-native') {
            devDependencies['@storybook/addon-ondevice-actions'] =
                versions_1.storybookReactNativeVersion;
            devDependencies['@storybook/addon-ondevice-backgrounds'] =
                versions_1.storybookReactNativeVersion;
            devDependencies['@storybook/addon-ondevice-controls'] =
                versions_1.storybookReactNativeVersion;
            devDependencies['@storybook/addon-ondevice-notes'] =
                versions_1.storybookReactNativeVersion;
        }
        if (schema.uiFramework.endsWith('-vite')) {
            if (!packageJson.dependencies['vite'] &&
                !packageJson.devDependencies['vite']) {
                devDependencies['vite'] = versions_1.viteVersion;
            }
        }
    }
    return (0, devkit_1.addDependenciesToPackageJson)(host, dependencies, devDependencies);
}
function addCacheableOperation(tree) {
    var _a;
    const nxJson = (0, devkit_1.readNxJson)(tree);
    if (!nxJson.tasksRunnerOptions ||
        !nxJson.tasksRunnerOptions.default ||
        (nxJson.tasksRunnerOptions.default.runner !==
            '@nx/workspace/tasks-runners/default' &&
            nxJson.tasksRunnerOptions.default.runner !== 'nx/tasks-runners/default')) {
        return;
    }
    nxJson.tasksRunnerOptions.default.options =
        nxJson.tasksRunnerOptions.default.options || {};
    nxJson.tasksRunnerOptions.default.options.cacheableOperations =
        nxJson.tasksRunnerOptions.default.options.cacheableOperations || [];
    if (!((_a = nxJson.tasksRunnerOptions.default.options.cacheableOperations) === null || _a === void 0 ? void 0 : _a.includes('build-storybook'))) {
        nxJson.tasksRunnerOptions.default.options.cacheableOperations.push('build-storybook');
    }
    (0, devkit_1.updateNxJson)(tree, nxJson);
}
function moveToDevDependencies(tree) {
    (0, devkit_1.updateJson)(tree, 'package.json', (packageJson) => {
        packageJson.dependencies = packageJson.dependencies || {};
        packageJson.devDependencies = packageJson.devDependencies || {};
        if (packageJson.dependencies['@nx/storybook']) {
            packageJson.devDependencies['@nx/storybook'] =
                packageJson.dependencies['@nx/storybook'];
            delete packageJson.dependencies['@nx/storybook'];
        }
        return packageJson;
    });
}
/**
 * This is a temporary fix for Storybook to support TypeScript configuration files.
 * The issue is that if there is a root tsconfig.json file, Storybook will use it, and
 * ignore the tsconfig.json file in the .storybook folder. This results in module being set
 * to esnext, and Storybook does not recognise the main.ts code as a module.
 */
function editRootTsConfig(tree) {
    if (tree.exists('tsconfig.json')) {
        (0, devkit_1.updateJson)(tree, 'tsconfig.json', (json) => {
            var _a;
            if (json['ts-node']) {
                json['ts-node'] = Object.assign(Object.assign({}, json['ts-node']), { compilerOptions: Object.assign(Object.assign({}, ((_a = json['ts-node'].compilerOptions) !== null && _a !== void 0 ? _a : {})), { module: 'commonjs' }) });
            }
            else {
                json['ts-node'] = {
                    compilerOptions: {
                        module: 'commonjs',
                    },
                };
            }
            return json;
        });
    }
}
function initGenerator(tree, schema) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const tasks = [];
        tasks.push(yield (0, js_1.initGenerator)(tree, Object.assign(Object.assign({}, schema), { skipFormat: true })));
        tasks.push(checkDependenciesInstalled(tree, schema));
        moveToDevDependencies(tree);
        editRootTsConfig(tree);
        addCacheableOperation(tree);
        return (0, devkit_1.runTasksInSerial)(...tasks);
    });
}
exports.initGenerator = initGenerator;
exports.default = initGenerator;
exports.initSchematic = (0, devkit_1.convertNxGenerator)(initGenerator);
